trieste/utils/pareto.py

```diff
@@ -106,6 +113,164 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+                dc,
+                lower,
+                upper,
+                number_of_objectives,
+                pf_ext,
+                pf_ext_idx,
+                jitter,
+                front,
+                total_size,
+                threshold,
+            ),
+            loop_vars=[dc, lower, upper],
+            shape_invariants=[
+                tf.TensorShape([None, 2, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+            ],
+        )
+        return BoundedVolumes(lower, upper)
+
+    def _while_body(
+        self,
+        dc,
+        lower,
+        upper,
+        number_of_objectives,
+        pf_ext,
+        pf_ext_idx,
+        jitter,
+        front,
+        total_size,
+        threshold,
+    ):
+        # Process test cell
+        dc = tf.unstack(dc, axis=0)
+        cell = dc[-1]
+        dc = tf.cond(
+            tf.not_equal(tf.size(dc[:-1]), 0),
+            lambda: tf.stack(dc[:-1]),
+            lambda: tf.zeros([0, 2, number_of_objectives], dtype=tf.int32),
+        )
+
+        arr = tf.range(number_of_objectives)
+        idx_lb = tf.gather_nd(pf_ext_idx, tf.stack((cell[0], arr), -1))
+        idx_ub = tf.gather_nd(pf_ext_idx, tf.stack((cell[1], arr), -1))
+        lb = tf.gather_nd(pf_ext, tf.stack((idx_lb, arr), -1))
+        ub = tf.gather_nd(pf_ext, tf.stack((idx_ub, arr), -1))
+
+        # Acceptance test:
+        test_accepted = self._is_test_required((ub - jitter) < front)
+        lower, upper = tf.cond(
+            test_accepted,
+            lambda: self._accepted_test_body(lower, upper, idx_lb, idx_ub),
+            lambda: (lower, upper),
+        )
+
+        # Reject test:
+        test_rejected = self._is_test_required((lb + jitter) < front)
+        dc = tf.cond(
+            tf.logical_and(test_rejected, tf.logical_not(test_accepted)),
+            lambda: self._rejected_test_body(cell, lb, ub, dc, total_size, threshold),
+            lambda: dc,
+        )
+
+        return dc, lower, upper
+
+    def _accepted_test_body(self, lower, upper, idx_lb, idx_ub):
+        lower = tf.concat([lower, idx_lb[None]], axis=0)
+        upper = tf.concat([upper, idx_ub[None]], axis=0)
+        return lower, upper
+
+    def _rejected_test_body(self, cell, lb, ub, dc, total_size, threshold):
+
+        dc_dist = cell[1] - cell[0]
+        hc_size = tf.math.reduce_prod(ub - lb, axis=0, keepdims=True)
+
+        # Only divide when it is not an unit cell
+        # and the volume is above the approx. threshold
+        not_unit_cell = tf.reduce_any(dc_dist > 1)
+        vol_above_thresh = tf.reduce_all((hc_size[0] / total_size) > threshold)
+        dc = tf.cond(
+            tf.logical_and(not_unit_cell, vol_above_thresh),
+            lambda: self._divide_body(dc, dc_dist, cell),
+            lambda: tf.identity(dc),
+        )
```

Comment:
was there a particular reason to use `tf.cond` over
```python
dc = self._divide_body(dc, dc_dist, cell) if tf.logical_and(not_unit_cell, vol_above_thresh) else tf.identity(dc)
```

Reply:
no actually, I'll update it

Reply:
Ah. I was thinking to replace all python if-else with `tf.cond`. it's not necessary here right?

Reply:
i might have said that. I can't remember. It's not important so we can leave it as is